### DISCUSSION TITLE
Sort columns

### DISCUSSION THREAD

**Discussion by jakubczajczynski** [nicolaasburgers/definition-library-releases#34] [Ideas]
Please add the ability to sort columns.

![356581704-bf7b8748-c5f2-4d6e-bece-cfe6545b311b](https://github.com/user-attachments/assets/13525b6e-7a44-45ea-9322-764abb1bc6b7)

**Comment by sheldon-walters** [comment]
Yes - exactly that

**Comment by jakubczajczynski** [comment]
I treat your plugin as a "quality of life" tool, so forgive me if my requests and suggestions are mundane. However, Definition Library is a digital "library" after all, which is why sorting and searching mechanisms seem so important to me.

Please, consider whether a better way to filter would be to use similar solutions that we know from Excel tables or file tables like in Windows Explorer, Directory Opus, or dozens of other programs. The plugin is developing in the right direction, but at the moment there are still many things that would improve the comfort of everyday use.

I don't know what GUI framework you are using and whether there are any ready-made solutions for more advanced tables for it. I suspect that because it is still a beta version, the GUI cannot be too refined yet, but also that the framework may not offer ready-made solutions that would be easy to implement, but a lot of work is needed here.

Some issues:

1. we cannot sort data in columns, e.g. ascending by date.

2. Columns do not remember their width after closing the window.

3. You cannot change the order of columns.

4. After closing the window, I would like the table filters not to reset and show the same state as before closing.

---

_A table with a filtering interface in the column headers would be more compact and easier to navigate._

![image](https://github.com/user-attachments/assets/7afe08ec-f52c-411d-93ea-8bb2242236fd)

_An intermediate solution would be to arrange the GUI elements used for filtering vertically, with the elements related to a given column being above the column._

![image](https://github.com/user-attachments/assets/195d843c-12ce-4aa1-a8bb-e72a70cceda1)

**Reply by nicolaasburgers** [reply]
Thanks again @jakubczajczynski for this detailed request, there are some great ideas in here.  Where would you position the "Originally Created In", "Names of Nested" and Clusterable/Hopable/File only aspects to the filtering?  Or do you personally not see the need for some of these at all?

**Reply by nicolaasburgers** [reply]
In the latest version - 0.3.2-beta - the columns are now sortable, and the column widths are remembered so that when you close and re-open, their widths should be as you had them the last time.  Let me konw what you think!

I've put being able to change the order of the columns lower down the list for now but I do like your suggestion of the placement of search fields above their relevant columns.  I'm just pondering whether that would require more room overall for the search controls, and less for the results list.

**Comment by jakubczajczynski** [comment]
I think "Names of Nested" and "Originally Created in" are very useful functions and should stay. I haven't used them yet because at the moment my clusters are not yet shared between different Grasshopper files. It's hard for me to say at the moment where these filters should be placed. It's definitely good to have them somewhere on the side, because it's a slightly different approach to displaying results.

Thank you for the enhancements in the columns section, it will make it easier to wait for bigger changes in the UI.

**Comment by jakubczajczynski** [comment]
I see you're using the month/day/year date format. My small, cosmetic wish would be to be able to present the date in the YYYY/MM/DD format - with leading zeros and maybe using a monospace font. Also, presenting the time in the 24h format would be nice. It's not important, but I admit that looking at this date format confuses me a bit. I hope this will not snowball into requesting different formats for different countries.

Date format apparently changed from what it was earlier, my previous posts shows DD.MM.YYYY.

In DefinitionLibrary 0.5 it looks like this:
![image](https://github.com/user-attachments/assets/97b69257-94d3-4a0c-b8bf-b513aa58d025)

**Reply by nicolaasburgers** [reply]
I actually don't set a specific date format - I just use what your system defines as the local time format and as far as I can see in my code, I haven't changed this since 0.3.x, so at this point I'm not sure how the dates could have changed for you but I'll have a look to see if I can add this as a new settting, where you toggle between DD/MM/YYYY, MM/DD/YYYY and YYYY-MM-DD formats.

**Reply by jakubczajczynski** [reply]
Ok, I see. This is probably something I can change on my end. Although my previous screenshots are from my personal computer (I see it's Win 11) and not my work computer (Win 10), my work computer is fully in Polish language and the Polish region and the dates are displayed as I would expect. My personal computer is different, the UI is in English but the region and dates seem to be set to Polish. Anyway, sorry for the confusion and this is something I should be able to configure on my end, so you don't have to bother.
